### DISCUSSION TITLE
SVRG

### DISCUSSION THREAD

**Discussion by ulricharmel** [google-deepmind/optax#226] [Q&A] [answered]
Hi,

I am not very family with optax. Can someone please, give me basic template of how I will implement Stochastic Variance Reduced Gradient
with Optax.

Thanks

**Comment by mkunesch** [comment, marked as answer]
Hi! Thanks for the question!

I'll answer for the equations of SVRG as outlined in [this](https://bingoh.github.io/2017/06/08/Stochastic-Variance-Reduced-Gradient-SVRG/) blogpost, i.e. the update rule:

w_t = w_{t−1}−η_t [∇ψ_{i_t}(w_{t−1})−∇ψ_{i_t} (w̃ )+∇P(w̃ )],

with variable names as in the blog post.

The design of optax is such that users calculate gradients and apply updates themselves (see #155 for reasons for the latter), so the best approach with optax would be to let the user calculate the three gradient terms in the square brackets, use optax gradient transforms to transform the gradients if necessary, and then write a custom `apply_updates` function that implements the equation above similar to `optax.apply_updates`. This function would be a great contribution to `update.py` in optax!

I hope this helps. Let me know if you have more questions!
